src/db/gc/mod.rs

```diff
@@ -0,0 +1,166 @@
+            depth,
+        );
+
+        while let Some(block) = stream.next().await {
+            let block = block?;
+            self.marked.remove(&truncated_hash(block.cid.hash()));
+        }
+
+        anyhow::Ok(())
+    }
+
+    // Remove marked keys from the database.
+    fn sweep(&mut self) -> anyhow::Result<()> {
+        let mut marked = HashSet::new();
+        mem::swap(&mut marked, &mut self.marked);
+        self.db.remove_keys(marked)
+    }
+
+    /// Starts the Garbage Collection loop.
+    ///
+    /// # Arguments
+    ///
+    /// * `depth` - Specifies how far back the full history should be maintained. Cannot be less
+    /// than chain finality.
+    /// * `interval` - GC Interval to avoid constantly consuming node's resources.
+    /// * `manual` - Whether or not the GC has to be triggered manually or automatically.
+    ///
+    /// NOTE: This currently does not take into account the fact that we might be starting the node
+    /// using CAR-backed storage with a snapshot, for implementation simplicity.
+    pub async fn gc_loop(&mut self, interval: Duration, manual: bool) -> anyhow::Result<()> {
+        loop {
+            match manual {
+                true => {
+                    let msg = self.gc_receiver.recv_async().await?;
+                    info!("running manual gc");
+                    let res = self.gc_workflow(manual).await;
+                    msg.send_async(res).await?;
+                    info!("finished manual gc run");
+                }
+                false => {
+                    self.gc_workflow(manual).await?;
+                    // Make sure we don't run the GC too often.
+                    time::sleep(interval).await;
+                }
+            }
+        }
+    }
+
+    async fn gc_workflow(&mut self, manual: bool) -> anyhow::Result<()> {
+        let depth = self.depth;
+        let tipset = self.chain_store.heaviest_tipset();
+        let current_epoch = tipset.epoch();
+        // Don't run the GC if there aren't enough state-roots yet. Manual GC overrides that.
+        if !manual || depth > current_epoch {
+            time::sleep(AVERAGE_BLOCK_TIME * (depth - current_epoch) as u32).await;
+            return anyhow::Ok(());
+        }
+
+        // Make sure we don't GC if we haven't advanced at least `depth` number of epochs since
+        // the last sweep. Manual GC overrides that.
+        let epochs_since_gc = current_epoch - self.epoch_sweeped;
+        if self.marked.is_empty() && !manual && epochs_since_gc < depth {
+            time::sleep(AVERAGE_BLOCK_TIME * epochs_since_gc as u32).await;
+            return anyhow::Ok(());
+        } else {
+            info!("populate keys for GC");
+            self.populate()?;
+        }
+
+        // Nothing to do.
+        if self.marked.is_empty() {
+            return anyhow::Ok(());
+        }
+
+        // Don't filter and sweep before we advance at least `depth`. Manual GC overrides that.
```

Comment:
Again, why does manually triggering the GC override these checks? It's a hard constraint that the GC may not remove any data generated within the last `chain_finality` epochs. This constraint must be followed regardless of how the GC is triggered.

Reply:
> Forks in the blockchain may occur in the last `chain_finality` epochs. The GC has to be aware of this. Write the module-level documentation, and then we can see if our mental models align.

Thank makes sense. As stated above, I think we have to get rid of the manual gc for good then. My initial implementation has taken this into account and the UX for such GC triggers is not amazing. We either have to change the GC cli to be `fire and forget` or get rid of it altogether.